### ISSUE TITLE
General MIDI Support?

### ISSUE THREAD

**Issue by OpenRift412** [opentyrian/opentyrian#47]
Is there any chance we could get support for General MIDI format for music playback? The FM Emulation is cool but it'd be nice to have more options.

**Comment by carlreinke** [comment]
Someone would need to reverse engineer how [Loudness](http://www.vgmpf.com/Wiki/index.php?title=LDS) maps to MIDI.  We don't have any source code for that.

**Comment by gobalius** [comment]
What if instead of using the LDS music, you integrate [FluidSynth](https://github.com/FluidSynth/fluidsynth) into OpenTyrian and use the LDS music converted to General MIDI files? [MIDI Player component](https://www.foobar2000.org/components/view/foo_midi) for [Foobar2000](https://www.foobar2000.org/) does support conversion of LDS to MIDI. I can also provide the converted files if needed. I hope that would not violate any licenses.

**Comment by gobalius** [comment]
I think there's an even more direct way, see https://github.com/stuerp/foo_midi/blob/main/3rdParty/libmidi/MIDIProcessorLDS.cpp

**Comment by nikitalita** [comment]
> Someone would need to reverse engineer how [Loudness](http://www.vgmpf.com/Wiki/index.php?title=LDS) maps to MIDI. We don't have any source code for that.

Do you no longer have access to the original source code? Or did you not have any MIDI related code in the production you got?

**Comment by carlreinke** [comment]
The Loudness source code was not included (although according to comments on [this page](https://ocremix.org/remix/OCR00972), the source code was made available to Simon Peter, the primary AdPlug developer).

**Comment by nikitalita** [comment]
Yeah, that makes sense, taking a look at the Loudness page on vgmpf. Well, taking a look at the raw recorded midi output from Tyrian in DOSBox, it seems that the only things that make it into the midi output from the LDS files are velocity, part volume, and panning. There's no mapping to CC messages to Tremolo, Portamento, etc. This is pretty much what libmidi does above. 

So, if we want to imitate the MIDI output from Tyrian, we could just copy that and call it a day. 

If we want to do something a bit more substantial, we'd have to devise a way to map the various soundbank settings (tremolo, portamento, etc.) and commands (like vibrate to modulate) to CC commands.

Also, Andras Molnar is on github. Maybe we can ask him nicely to release the source for Loudness? 😇

**Comment by gobalius** [comment]
> I think there's an even more direct way, see https://github.com/stuerp/foo_midi/blob/main/3rdParty/libmidi/MIDIProcessorLDS.cpp

excuse me @carlreinke and @nikitalita - does this what I linked to not help at all? IMHO that's exactly the mapping you are talking about, isn't it?

**Comment by nikitalita** [comment]
I have it working. I'll PR it soon.
